src/module/service.cpp

```diff
@@ -0,0 +1,290 @@
+#include <grpcpp/server_context.h>
+#include <grpcpp/support/status.h>
+#include <module/v1/module.grpc.pb.h>
+#include <module/v1/module.pb.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+
+#include <boost/log/trivial.hpp>
+#include <boost/none.hpp>
+#include <common/utils.hpp>
+#include <components/component_base.hpp>
+#include <components/generic.hpp>
+#include <components/service_base.hpp>
+#include <config/resource.hpp>
+#include <memory>
+#include <module/handler_map.hpp>
+#include <module/service.hpp>
+#include <registry/registry.hpp>
+#include <resource/resource.hpp>
+#include <resource/resource_base.hpp>
+#include <robot/client.hpp>
+#include <robot/service.hpp>
+#include <rpc/server.hpp>
+#include <string>
+#include <subtype/subtype.hpp>
+
+namespace {
+Dependencies get_dependencies(ModuleService_* m,
+                              google::protobuf::RepeatedPtrField<std::string> proto) {
+    Dependencies deps;
+    for (auto& dep : proto) {
+        auto name = Name::from_string(dep);
+        std::shared_ptr<ResourceBase> resource = m->get_parent_resource(name);
+        deps.emplace(name, resource);
+    }
+    return deps;
+}
+}  // namespace
+   //
+std::shared_ptr<ResourceBase> ModuleService_::get_parent_resource(Name name) {
+    if (parent == nullptr) {
+        parent = RobotClient::at_address("unix://" + parent_addr, {0, boost::none});
+    }
+
+    return parent->resource_by_name(name.to_proto());
+}
+
+::grpc::Status ModuleService_::AddResource(::grpc::ServerContext* context,
+                                           const ::viam::module::v1::AddResourceRequest* request,
+                                           ::viam::module::v1::AddResourceResponse* response) {
+    viam::app::v1::ComponentConfig proto = request->config();
+    Component cfg = Component::from_proto(proto);
+    std::shared_ptr<Module> module = this->module;
+    module->lock.lock();
+
+    std::shared_ptr<ResourceBase> res;
+    if (cfg.api.is_component_type()) {
+        std::shared_ptr<ComponentRegistration> reg = Registry::lookup_component(cfg.api, cfg.model);
+        if (reg != nullptr) {
+            res = reg->create_rpc_client(module->name, module->channel);
+        };
+    } else if (cfg.api.is_service_type()) {
+        std::shared_ptr<ServiceRegistration> reg = Registry::lookup_service(cfg.api, cfg.model);
+        if (reg != nullptr) {
+            res = reg->create_rpc_client(module->name, module->channel);
+        }
+    } else {
+        module->lock.unlock();
+        return grpc::Status(grpc::UNKNOWN, "unknown resource type " + cfg.api.resource_type);
+    }
+
+    if (module->services.find(cfg.api) == module->services.end()) {
+        module->lock.unlock();
+        return grpc::Status(grpc::UNKNOWN, "module cannot service api " + cfg.api.to_string());
+    }
+
+    std::shared_ptr<SubtypeService> sub_svc = module->services.at(cfg.api);
+    if (cfg.api.is_component_type() && !(cfg.api == Generic::subtype())) {
```

Comment:
given the DoCommand project, would just omit this entirely

Reply:
handling in subsequent pr.